### PR TITLE
feat(db): CreateUndeployVersion and Undeploy transformers migrated to DB

### PR DESCRIPTION
**Pull request by miguel-crespo-fdc**
Ref: SRX-0E8EKB

### CODE REVIEW

File: pkg/db/overview.go [hunk 1]
```diff
@@ -147,7 +147,9 @@ func (h *DBHandler) UpdateOverviewDeploymentAttempt(ctx context.Context, transac
 	if app == nil {
 		return fmt.Errorf("could not find application %s in environment %s in overview", queuedDeployment.App, queuedDeployment.Env)
 	}
-	app.QueuedVersion = uint64(*queuedDeployment.Version)
+	if queuedDeployment.Version != nil {
+		app.QueuedVersion = uint64(*queuedDeployment.Version)
+	}
```
Comment (miguel-crespo-fdc) [inline pkg/db/overview.go:152]: Queued deployment might be nil

File: services/cd-service/pkg/repository/transformer.go [hunk 2]
```diff
@@ -423,6 +423,9 @@ func (s *State) GetLastRelease(ctx context.Context, transaction *sql.Tx, fs bill
 		if err != nil {
 			return 0, fmt.Errorf("could not get releases of app %s: %v", application, err)
 		}
+		if releases == nil || len(releases.Metadata.Releases) == 0 {
+			return 0, nil
+		}
 		l := len(releases.Metadata.Releases)
 		return uint64(releases.Metadata.Releases[l-1]), nil
```
Comment (miguel-crespo-fdc) [inline services/cd-service/pkg/repository/transformer.go:430]: Guard against empty releases (there was no guard for l = 0)

File: services/cd-service/pkg/repository/transformer.go [hunk 3]
```diff
@@ -609,13 +627,13 @@ func (c *CreateApplicationVersion) Transform(
 	sortedKeys := sorting.SortKeys(c.Manifests)
 
 	if state.DBHandler.ShouldUseOtherTables() {
-		anyRelease, err := state.DBHandler.DBSelectAnyRelease(ctx, transaction)
+		anyRelease, err := state.DBHandler.DBSelectReleasesByApp(ctx, transaction, c.Application, false)
 		if err != nil {
```
Comment (miguel-crespo-fdc) [inline services/cd-service/pkg/repository/transformer.go:631]: DBSelectAnyRelease Did not make any sense here? Any release simply checks for existence of any release for any app on any env, and we then make decisions based on the result, which cause problems with previously deleted applications

Comment (miguel-crespo-fdc) [reply services/cd-service/pkg/repository/transformer.go:631]: And I don't know how it didn't come up earlier

File: pkg/db/db.go [hunk 4]
```diff
@@ -1404,6 +1428,54 @@ func (h *DBHandler) DBSelectDeployment(ctx context.Context, tx *sql.Tx, appSelec
...
+		result = append(result, *row)
+	}
+	err = rows.Close()
+	if err != nil {
+		return nil, fmt.Errorf("deployments: row closing error: %v\n", err)
+	}
+	err = rows.Err()
+	if err != nil {
+		return nil, fmt.Errorf("deployments: row has error: %v\n", err)
+	}
```
Comment (sven-urbanski-freiheit-com) [inline pkg/db/db.go:1475]: please use `CloseRows(row)` instead

File: services/cd-service/pkg/repository/repository.go [hunk 5]
```diff
@@ -2601,7 +2601,16 @@ func extractPrNumber(sourceMessage string) string {
 	}
 }
 
-func (s *State) IsUndeployVersion(application string, version uint64) (bool, error) {
+func (s *State) IsUndeployVersion(ctx context.Context, transaction *sql.Tx, application string, version uint64) (bool, error) {
+	if s.DBHandler.ShouldUseOtherTables() && transaction != nil {
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/repository.go:2605]:
```suggestion
	if s.DBHandler.ShouldUseOtherTables() {
```

Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/repository/repository.go:2605]: I don't think we need this extra check.

File: services/cd-service/pkg/repository/transformer.go [hunk 6]
```diff
@@ -1251,86 +1347,84 @@ func (u *UndeployApplication) Transform(
...
+
+			var isUndeploy bool
+			if deployment != nil && deployment.Version != nil {
+				release, err := state.DBHandler.DBSelectReleaseByVersion(ctx, transaction, u.Application, uint64(*deployment.Version))
+				if err != nil {
+					return "", err
+				}
+				isUndeploy = release.Metadata.UndeployVersion
+				if !isUndeploy {
+					return "", fmt.Errorf("UndeployApplication(db): error cannot un-deploy application '%v' the release '%v' is not un-deployed", u.Application, env)
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer.go:1364]:
```suggestion
					return "", fmt.Errorf("UndeployApplication(db): error cannot un-deploy application '%v' the current release '%v' is not un-deployed", u.Application, env)
```

File: services/manifest-repo-export-service/pkg/cmd/server.go [hunk 7]
```diff
@@ -398,6 +398,12 @@ func getTransformer(ctx context.Context, eslEventType db.EventType) (repository.
...
 		//exhaustruct:ignore
 		return &repository.DeleteEnvFromApp{}, nil
+	case db.EvtCreateUndeployApplicationVersion:
+		//exhaustruct:ignore
+		return &repository.CreateUndeployApplicationVersion{}, nil
+	case db.EvtUndeployApplication:
+		//exhaustruct:ignore
+		return &repository.UndeployApplication{}, nil
 	default:
 		logger.FromContext(ctx).Sugar().Warnf("Found an unknown event %s. No further events will be processed.", eslEventType)
```
Comment (sven-urbanski-freiheit-com) [inline services/manifest-repo-export-service/pkg/cmd/server.go:408]: I think this was the last transformer to be migrated, right?
In that case, the `default` block here should return an error and not just log a warning.

File: services/manifest-repo-export-service/pkg/repository/repository.go [hunk 8]
```diff
@@ -974,6 +974,30 @@ func (s *State) GetEnvironmentLocksFromDB(ctx context.Context, transaction *sql.
...
+	if err != nil {
+		return 0, err
+	}
+	if entries, err := fs.ReadDir(releasesDir); err != nil {
+		return 0, err
+	} else {
+		var lastRelease uint64 = 0
+		for _, e := range entries {
+			if i, err := strconv.ParseUint(e.Name(), 10, 64); err != nil {
+				//TODO(HVG): decide what to do with bad named releases
```
Comment (sven-urbanski-freiheit-com) [inline services/manifest-repo-export-service/pkg/repository/repository.go:990]: let's just log this as a warning or error, instead of leaving the TODO in.

**Review by sven-urbanski-freiheit-com — CHANGES_REQUESTED** [review]
see comments about DBSelectReleasesByApp and default in export-service

File: services/manifest-repo-export-service/pkg/cmd/server.go [hunk 9]
```diff
@@ -398,10 +398,14 @@ func getTransformer(ctx context.Context, eslEventType db.EventType) (repository.
...
-		logger.FromContext(ctx).Sugar().Warnf("Found an unknown event %s. No further events will be processed.", eslEventType)
-		return nil, nil
+	case db.EvtCreateUndeployApplicationVersion:
+		//exhaustruct:ignore
+		return &repository.CreateUndeployApplicationVersion{}, nil
+	case db.EvtUndeployApplication:
+		//exhaustruct:ignore
+		return &repository.UndeployApplication{}, nil
 	}
+	return nil, nil
```
Comment (sven-urbanski-freiheit-com) [inline services/manifest-repo-export-service/pkg/cmd/server.go:408]:
```suggestion
	return nil, fmt.Errorf("Could not find transformer for event type %v", eslEventType)
```

**Review by sven-urbanski-freiheit-com — APPROVED** [review]

**Pull request merged**